File: packages/silo-client-java/src/main/java/in/org/quicko/silo/client/cache/ResponseCache.java
package in.org.quicko.silo.client.cache;

import com.fasterxml.jackson.databind.JsonNode;
import com.github.benmanes.caffeine.cache.Caffeine;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.function.Supplier;

/**
 * The responses one {@code Transport} is holding, as one Caffeine cache per
 * {@link CachePolicy} in force.
 *
 * <p>A write invalidates by path prefix and not by key, because the caller that
 * replaced one entry cannot name the pages that entry appeared on.
 *
 * <p>Held per transport, so {@code withKey} and {@code withUrl} start empty:
 * one key's reads are not another key's to serve.
 */
public final class ResponseCache {
  private final CacheOptions options;
  private final Map<CachePolicy, com.github.benmanes.caffeine.cache.Cache<CacheKey, JsonNode>>
      caches = new ConcurrentHashMap<>();

  public ResponseCache(CacheOptions options) {
    this.options = options == null ? CacheOptions.off() : options;
  }

  public boolean isEnabled() {
    return options.enabled();
  }

  /**
   * The stored response, or what {@code load} answers. A load that throws stores
   * nothing, so a failed read is never remembered.
   *
   * <p>Read then write, rather than Caffeine's loading {@code get}: that one
   * computes inside {@code ConcurrentHashMap.compute} and would hold a bin lock
   * across the request. The cost is that two concurrent misses both fetch.
   */
  public JsonNode get(
      CachePolicy declared,
      String method,
      String path,
      Map<String, Object> query,
      Supplier<JsonNode> load) {
    CachePolicy policy = options.inForce(declared);
    if (policy == null) return load.get();

    CacheKey key = CacheKey.of(method, path, query);
    com.github.benmanes.caffeine.cache.Cache<CacheKey, JsonNode> held = held(policy);
    JsonNode stored = held.getIfPresent(key);
    if (stored != null) return stored;

    JsonNode fetched = load.get();
    held.put(key, fetched);
    return fetched;
  }

  /** Drops every stored response addressed at, or below, this path. */
  public void invalidate(String pathPrefix) {
    if (pathPrefix == null) return;
    for (com.github.benmanes.caffeine.cache.Cache<CacheKey, JsonNode> cache : caches.values()) {
      cache.asMap().keySet().removeIf(stored -> stored.matches(pathPrefix));
    }
  }

  /** Drops everything — the escape hatch after a write this client did not make. */
  public void clear() {
    caches.values().forEach(com.github.benmanes.caffeine.cache.Cache::invalidateAll);
  }

  public CacheStatistics statistics() {
    CacheStatistics total = CacheStatistics.Empty;
    for (com.github.benmanes.caffeine.cache.Cache<CacheKey, JsonNode> cache : caches.values()) {
      com.github.benmanes.caffeine.cache.stats.CacheStats stats = cache.stats();
      total = total.plus(new CacheStatistics(
          stats.hitCount(), stats.missCount(), stats.evictionCount(), cache.estimatedSize()));
    }
    return total;
  }

  private com.github.benmanes.caffeine.cache.Cache<CacheKey, JsonNode> held(CachePolicy policy) {
    return caches.computeIfAbsent(policy, this::build);
  }

  private com.github.benmanes.caffeine.cache.Cache<CacheKey, JsonNode> build(CachePolicy policy) {
    Caffeine<Object, Object> builder = Caffeine.newBuilder()
        .expireAfterWrite(policy.ttl())
        .maximumSize(policy.maxSize())
        .recordStats();
    if (options.ticker() != null) builder.ticker(options.ticker());
    return builder.build();
  }
}

File: packages/silo-client-dart/lib/src/cache/response_cache.dart
import 'package:clock/clock.dart';

import 'cache_key.dart';
import 'cache_options.dart';
import 'cache_policy.dart';
import 'cache_statistics.dart';

typedef _Stored = ({String body, DateTime storedAt});

/// The JSON one client holds, in one store per ttl and bound in force. A store
/// is insertion-ordered, so its oldest entry is both next to expire and next
/// to evict. Time comes from `package:clock`, so tests use `withClock`.
final class ResponseCache {
  ResponseCache(this._options);

  final CacheOptions _options;
  final Map<(Duration, int), Map<String, _Stored>> _stores = {};
  int _hits = 0;
  int _misses = 0;
  int _evictions = 0;

  bool get isEnabled => _options.enabled;

  CacheStatistics get statistics =>
      CacheStatistics(_hits, _misses, _evictions, _stores.values.fold(0, (size, store) => size + store.length));

  /// The stored body, or what [load] answers. A failed load stores nothing,
  /// and two concurrent misses both load.
  Future<String?> get(
    CachePolicy declared,
    String method,
    String path,
    Map<String, Object?> query,
    Future<String?> Function() load,
  ) async {
    if (!isEnabled) return load();

    final (ttl, maxSize) = _inForce(declared);
    final store = _stores.putIfAbsent((ttl, maxSize), () => {});
    _expire(store, ttl);

    final key = CacheKey.of(method, path, query);
    if (store[key] case final stored?) {
      _hits++;
      return stored.body;
    }

    _misses++;
    final fetched = await load();
    if (fetched != null) {
      store.remove(key);
      while (store.length >= maxSize) {
        store.remove(store.keys.first);
        _evictions++;
      }
      store[key] = (body: fetched, storedAt: clock.now());
    }
    return fetched;
  }

  /// Drops everything stored at or below [pathPrefix].
  void invalidate(String? pathPrefix) {
    if (pathPrefix == null) return;
    for (final store in _stores.values) {
      store.removeWhere((key, _) => CacheKey.matches(key, pathPrefix));
    }
  }

  /// The escape hatch after a write this client did not make.
  void clear() {
    for (final store in _stores.values) {
      store.clear();
    }
  }

  (Duration, int) _inForce(CachePolicy declared) {
    final ttl = declared.ttl ?? _options.ttl;
    final maxSize = declared.maxSize ?? _options.maxSize;
    if (ttl == null || maxSize == null) {
      throw StateError('a cached read needs a ttl and a maxSize, on its CachePolicy or on CacheOptions');
    }
    if (ttl <= Duration.zero || maxSize <= 0) {
      throw ArgumentError('a cache ttl and maxSize have to be positive');
    }
    return (ttl, maxSize);
  }

  void _expire(Map<String, _Stored> store, Duration ttl) {
    final now = clock.now();
    while (store.isNotEmpty && now.difference(store.values.first.storedAt) >= ttl) {
      store.remove(store.keys.first);
      _evictions++;
    }
  }
}
